openedx/enterprise-catalog
# fix: conditionally update json_metadata for courses with specific fields from /search/all data

## Description

The first stage of our enterprise catalog jobs are to source data from discovery's /search/all endpoint. We previously updated/created ContentMetadata object's `json_metadata` field with the metadata returned by this /search/all endpoint. However, we rely on the full course metadata for our ContentMetadata objects with content_type == "course" for Algolia indexing. This is metadata is populated via `update_full_content_metadata_task`. 

Due to our reliance on the full course metadata, we no longer want to fully replace the `json_metadata` field of existing ContentMetadata records with the /search/all metadata as we have seen several cases where the `update_content_metadata` Jenkins job times out, and the full course metadata update never occurs. This means the Algolia indexing that occurs an hour or two later does not have the necessary data for Algolia re-indexing, causing partner logo and name to be removed from our search results in the Learner Portal and elsewhere.

This refactoring also switches to using a bulk DB operation (i.e., `bulk_update`) for updating existing ContentMetadata objects and an atomic DB transaction for creating new ContentMetadata objects, which should improve task performance.

## Ticket Link

[ENT-4208](https://openedx.atlassian.net/browse/ENT-4208)

## Post-review

Squash commits into discrete sets of changes

## Testing instructions

1. Ensure no ContentMetadata records exist in your devstack database, you have existing enterprise catalog(s) for course content, and your discovery service is running.
2. Run `./manage.py update_content_metadata` in the enterprise-catalog shell.
3. Observe ContentMetadata records having been created using the `/search/all` metadata. For courses, expect to see the merged `/search/all` metadata with the full course metadata.
4. Edit the `json_metadata` for an existing course to remove fields that only exist in the `/search/all` metadata (e.g., "aggregation_key", and re-run `./manage.py update_content_metadata` (you may need to temporarily comment out the `@expiring_task_semaphore` decorator on the tasks).
5. Observe that despite removing "aggregation_key" from an existing ContentMetadata, the `json_metadata` is populated again with the plucked fields, including "aggregation_key", from the `/search/all` metadata.
